src/charm.py

```diff
@@ -116,14 +138,33 @@ def _mysql_layer(self, service_name: str) -> Layer:
     # ---- Properties
     @property
     def _config(self):
-        """Configuration data for MySQL authentication."""
+        """Configuration data for MySQL."""
         config = {
+            "innodb_cluster_name": self.model.config["innodb_cluster_name"],
             "mysql_user": "mysql_user",
-            "mysql_password": self._generate_password(),
-            "mysql_root_password": self._generate_password(),
+            "mysql_password": self._mysql_password,
+            "mysql_root_password": self._root_password,
         }
         return config
 
+    @property
+    def _root_password(self):
+        root_password = self._stored.root_password
```

Comment:
I like this encapsulation! But you should get the root_password from the config, rather than the StoredState.

Reply:
Hmm, I think probably not in this case. Having passwords in config makes them a little tooooo easy to reveal.

A common pattern is autogenerate a root pw, and have an action to retrieve that initial password (generally stored on the peer relation) 

(We need Juju secrets...!)

Reply:
The password is already stored in config in this draft. I agree that, without a released implementation of Juju Secrets, this is not a great practice.

Having the leader generate and share the password is a more secure interim practice. We'd only want to cache the password's value in StoredState if the value is unique to reach unit, and we don't mind it changing whenever the pod is rescheduled.